Add iceoryx2 UTransport specification (#285)
eclipse-uprotocol/up-spec: up-l1/iceoryx2.adoc

--
All types of uProtocol messages *MUST* be transferred using iceoryx2's PublishSubscribe MessagingPattern only. iceoryx2's RequestResponse API *MUST NOT* be used.
--

== UMessage Mapping

A uProtocol message consists of _UAttributes_ and optional payload. The following sections define how these are mapped to/from iceoryx2 Services.

=== UAttributes

[.specitem,oft-sid="dsn~up-transport-iceoryx2-attributes-mapping~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--
The value of a iceoryx2 Service that is used to convey a uProtocol message *MUST* have a https://docs.rs/iceoryx2/0.6.1/iceoryx2/service/builder/publish_subscribe/struct.Builder.html#method.user_header[user header] as defined in <<uAttributes Mapping to iceoryx2 user header>>.
--

==== uAttributes Mapping to iceoryx2 user header

[source,rust]
----
const MAX_FEASIBLE_UATTRIBUTES_SERIALIZED_LENGTH: usize = 1000; // choosing ~1000 u8s
                                                                // somewhat arbitrarily
                                                                // this should be confirmed

#[repr(C)]
#[derive(ZeroCopySend)]
struct UProtocolHeader {

>     uprotocol_schema_version: u8,

Reviewer: ```suggestion
    uprotocol_major_version: u8,
```

Reviewer: There is no place in the uProtocol specification where we use the term _uProtocol Schema Version_, if I am not mistaken. Is the idea here that a receiver that has been built with uProtocol major version 2 is still able to process a message that has been created by sender based on uProtocol major version 1?

Author: Ah, I'm misremembering the Zenoh UTransport spec. You're right!

    uattributes_serialized: FixedSizeVec<u8, MAX_FEASIBLE_UATTRIBUTES_SERIALIZED_LENGTH>
}
----

.uAttributes Mapping to UProtocolHeader, iceoryx2 User Header
[%autowidth]
|===
| field | value

> | `uprotocol_schema_version`

Reviewer: ```suggestion
| `uprotocol_major_version`
```

Author: Same comment as I left above.

a| *MUST* be set to the bytes representing the protobuf encoding of the UAttributes object as defined by the link:../up-core-api/uprotocol/v1/uattributes.proto[UAttributes Protocol Buffers definition]

|===

_Known limitations_: Using a `FixedSizeVec` in the header containing a serialization of the entire `UAttributes` is wasteful and unnecessarily makes every message larger than it really need be most of the time. The approach chosen here was chosen primarily for expediency and simplicity of implementation. It should be possible to optimize this significantly. Some low-hanging fruit: use custom structs for entirety of `UAttributes` in User Header, make use of `Option<T>` to cut down on size when e.g. not using token or traceparent.

==== Message Priority

Currently not a clear path to map from uProtocol message priority to iceoryx2 QoS (Quality of Service). Skip for current revision.

=== Payload Mapping

[.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-mapping~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--
An iceoryx2 message that is used to convey a uProtocol message *MUST* contain in its payload the unaltered value of the UMessage's _payload_ field.
--

> [.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-mapping~2",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]

Reviewer: ```suggestion
[.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-zero-copy~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
```

this needs to get a different name, the number at the end of the ID is the *revision* of the specitem ...

--
An iceoryx2 message that is used to convey a uProtocol message *MUST* be zero-copy and shared-memory compatible (i.e. be fixed in size, contain no pointers, and so on).
--

== iceoryx2 Service Name Structure

> Message producers publish messages to the iceory2 inter-process communication using _ServiceNames_. Other clients can indicate their interest in particular patterns in order to receive the messages that are being published using matching ServiceNames.

Reviewer: ```suggestion
Message producers publish messages to the iceory2 inter-process communication using _ServiceNames_. Other clients can then use the same _ServiceName_ to receive the messages that are being published.
```

The iceoryx2 _ServiceName_ that is used to transfer a uProtocol message is derived from the message's `source` and `sink` attributes.

[.specitem,oft-sid="dsn~up-transport-iceoryx2-service-name~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--

> The ServiceName of an iceoryx2 message containing a _Publish_ UMessage **MUST** consist of the following segments:

Reviewer: ```suggestion
The ServiceName for sending a _Publish_ UMessage **MUST** consist of the following segments:
```

`up/[source.authority]/[source.ue_type]/[source.ue_instance]/[source.ue_version]/[source.resource]`

> The ServiceName of an iceoryx2 message containing a _Notification_ or _RPC Response_ UMessage **MUST** consist of the following segments:

Reviewer: ```suggestion
The ServiceName for sending a _Notification_ or _RPC Response_ UMessage **MUST** consist of the following segments:
```

`up/[source.authority]/[source.ue_type]/[source.ue_instance]/[source.ue_version]/[source.resource]/[sink.authority]/[sink.ue_type]/[sink.ue_instance]/[sink.ue_version]/[sink.resource]`

> The ServiceName of an iceoryx2 message containing an _RPC Request_ UMessage **MUST** consist of the following segments:

Reviewer: ```suggestion
The ServiceName for sending an _RPC Request_ UMessage **MUST** consist of the following segments:
```

`up/[sink.authority]/[sink.ue_type]/[sink.ue_instance]/[sink.ue_version]/[sink.resource]`

Please refer to <<UUri Encoding Rules>> for details regarding the encoding of the `source` and `sink` UUris into the key expression's segments.

--

=== UUri Encoding Rules

The table below contains the rules for encoding a UUri's fields into an iceoryx2 ServiceName's segments.

[cols="2,2,6"]
|===
| ServiceName Segment
| UUri Field
| Encoding

|`authority`
|`authority_name`
a| The segment *MUST* contain the (UTF8) string representation of the 

1. name of the host/authority that the (local) uEntity is running on, if authority name is empty.
2. authority name, otherwise.

|`ue_type`
|`ue_id`

> a| The segment *MUST* contain the (UTF8) string representation of the
> 
> the upper-case link:https://www.rfc-editor.org/rfc/rfc4648#section-8[base16 encoding] of the uEntity type identifier with all leading `0` characters omitted.

Reviewer: ```suggestion
a| The segment *MUST* contain the (UTF8) string representation of the
upper-case link:https://www.rfc-editor.org/rfc/rfc4648#section-8[base16 encoding] of the uEntity type identifier with all leading `0` characters omitted.
```

|`ue_instance`
|`ue_id`

> a| The segment *MUST* contain the (UTF8) string representation of the
> 
> the upper-case link:https://www.rfc-editor.org/rfc/rfc4648#section-8[base16 encoding] of the uEntity instance identifier with all leading `0` characters omitted.

Reviewer: ```suggestion
a| The segment *MUST* contain the (UTF8) string representation of the
upper-case link:https://www.rfc-editor.org/rfc/rfc4648#section-8[base16 encoding] of the uEntity instance identifier with all leading `0` characters omitted.
```

|`ue_version`
|`ue_version_major`
a| The segment *MUST* contain the (UTF8) string representation of the

the upper-case link:https://www.rfc-editor.org/rfc/rfc4648#section-8[base16 encoding] of the uEntity major version with all leading `0` characters omitted.

|`resource`
|`resource_id`

> a| The segment *MUST* contain the (UTF8) string representation of the
> 
> the upper-case link:https://www.rfc-editor.org/rfc/rfc4648#section-8[base16 encoding] of the resource identifier with all leading `0` characters omitted.

Reviewer: ```suggestion
a| The segment *MUST* contain the (UTF8) string representation of the
upper-case link:https://www.rfc-editor.org/rfc/rfc4648#section-8[base16 encoding] of the resource identifier with all leading `0` characters omitted.
```

=== UAttributes

[.specitem,oft-sid="dsn~up-transport-iceoryx2-attributes-mapping~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--
The value of a iceoryx2 Service that is used to convey a uProtocol message *MUST* have a https://docs.rs/iceoryx2/0.6.1/iceoryx2/service/builder/publish_subscribe/struct.Builder.html#method.user_header[user header] as defined in <<uAttributes Mapping to iceoryx2 user header>>.
--

==== uAttributes Mapping to iceoryx2 user header

[source,rust]
----
const MAX_FEASIBLE_UATTRIBUTES_SERIALIZED_LENGTH: usize = 1000; // choosing ~1000 u8s
                                                                // somewhat arbitrarily
                                                                // this should be confirmed

#[repr(C)]
#[derive(ZeroCopySend)]
struct UProtocolHeader {
    uprotocol_major_version: u8,
    uattributes_serialized: FixedSizeVec<u8, MAX_FEASIBLE_UATTRIBUTES_SERIALIZED_LENGTH>
}
----

.uAttributes Mapping to UProtocolHeader, iceoryx2 User Header
[%autowidth]
|===
| field | value

| `uprotocol_major_version`

> a| *MUST* be set to a single byte representing the UProtocol major version (`0x01` at the time of writing)

Author: ```suggestion
a| *MUST* be set to a single byte representing the uProtocol major version (`0x01` at the time of writing)
```

==== Message Priority

Currently not a clear path to map from uProtocol message priority to iceoryx2 QoS (Quality of Service). Skip for current revision.

=== Payload Mapping

[.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-mapping~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--
An iceoryx2 message that is used to convey a uProtocol message *MUST* contain in its payload the unaltered value of the UMessage's _payload_ field.
--

[.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-zero-copy~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--
An iceoryx2 message that is used to convey a uProtocol message *MUST* be zero-copy and shared-memory compatible:

* is self contained, no heap, no pointers to external sources, no handles to external resources (files etc)
* has a uniform memory representation -> `#[reprpass:[(C)]]`
* does not use pointers to manage their internal structure
* and its members don't implement `Drop` explicitly

> * has a `'static` lifetime

Reviewer: these are all criteria that are defined based on Rust concepts, right? In the next section you then refer to the corresponding Trait _if implemented in Rust_. To me it seems that the only way to implement this transport is indeed using Rust, or am I mistaken? If so, then we could simply refer to the Trait and leave out the specific requirements. That way, we would be less prone to running into inconsistencies when iceoryx2 evolves ...

Author: Hmm... you're right that this is mixed up a bit.

It is possible to have a C or C++ payload meet the criteria as well, as we can see in these [examples](https://github.com/eclipse-iceoryx/iceoryx2/tree/main/examples/c/publish_subscribe_cross_language/src).

Let me try to more clearly separate this out.

Author: Could you take a look at [e220df9](https://github.com/eclipse-uprotocol/up-spec/pull/285/commits/e220df9116c090f128726e12d4042656abae4934) to see if this helps clear up the separation between general requirements on the payload and Rust-specific?

a| *MUST* be set to a single byte representing the uProtocol major version (`0x01` at the time of writing)

| `uattributes_serialized`
a| *MUST* be set to the bytes representing the protobuf encoding of the UAttributes object as defined by the link:../up-core-api/uprotocol/v1/uattributes.proto[UAttributes Protocol Buffers definition]

|===

_Known limitations_: Using a `FixedSizeVec` in the header containing a serialization of the entire `UAttributes` is wasteful and unnecessarily makes every message larger than it really need be most of the time. The approach chosen here was chosen primarily for expediency and simplicity of implementation. It should be possible to optimize this significantly. Some low-hanging fruit: use custom structs for entirety of `UAttributes` in User Header, make use of `Option<T>` to cut down on size when e.g. not using `token` or `traceparent`.

==== Message Priority

Currently not a clear path to map from uProtocol message priority to iceoryx2 QoS (Quality of Service). Skip for current revision.

=== Payload Mapping

[.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-mapping~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--

> An iceoryx2 message that is used to convey a uProtocol message *MUST* contain in its payload the unaltered value of the UMessage's _payload_ field.

Reviewer: ```suggestion
An iceoryx2 `Sample` that is used to convey a uProtocol message *MUST* contain in its payload the unaltered value of the UMessage's _payload_ field.
```

--

[.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-zero-copy~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--
An iceoryx2 message that is used to convey a uProtocol message *MUST* be zero-copy and shared-memory compatible:

* is self contained, no heap, no pointers to external sources, no handles to external resources (files etc)
* has a uniform memory representation
* does not use pointers to manage their internal structure
* has a lifetime of the entire program duration
--

The cleanest way to describe this is to say that the payload definition **MUST** have the link:https://docs.rs/iceoryx2/0.6.1/iceoryx2/prelude/trait.ZeroCopySend.html[`ZeroCopySend`] trait implemented for it if written in Rust.

That encapsulates the following additional properties in Rust:

* Rust `enum` and `struct` are marked  as `#[reprpass:[(C)]]`
* members don't implement `Drop` explicitly

> * has a `'static` lifetime

Reviewer: I wonder if this is actually needed at all, given that this is already enforced by iceoryx2 itself ...

Author: Okay, let's keep this wording for now.

See the NOTICE file(s) distributed with this work for additional
information regarding copyright ownership.

This program and the accompanying materials are made available under
the terms of the Apache License Version 2.0 which is available at
https://www.apache.org/licenses/LICENSE-2.0
 
SPDX-FileType: DOCUMENTATION
SPDX-License-Identifier: Apache-2.0
----

== Overview

https://iceoryx.io[Eclipse iceoryx2&trade;] is an efficient, and ultra-low latency inter-process communication middleware. The library is designed to provide you with fast and reliable zero-copy and lock-free inter-process communication mechanisms.

This document defines how uProtocol messages can be exchanged by means of iceoryx2 _Services_.

== iceoryx2 Version

[.specitem,oft-sid="dsn~up-transport-iceoryx2-protocol-version~1",oft-needs="impl",oft-tags="TransportLayerImpl"]
--
Each transport implementing this specification **MUST** use version `0.6.1` of the iceoryx2 protocol in order to ensure interoperability of different language libraries.
--

== MessagingPattern

[.specitem,oft-sid="dsn~up-transport-iceoryx2-messaging-pattern~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--

> All types of uProtocol messages *MUST* be transferred using iceoryx2's PublishSubscribe MessagingPattern only. iceoryx2's RequestResponse API *MUST NOT* be used.

Reviewer 2: @sophokles73 It would introduce a syscall on the sender side, which would cause latency, and when the receiver never acquires the notifications/samples the underlying buffer of the event mechanism might run out of memory, which the user has to handle - but this depends on the mechanism the user would like to use for events. iceoryx2 provides here semaphores (which do not have the problem but are not interoperable with file descriptors in the context of an event multiplexer like `select` or `epoll`) and unix datagram sockets that are interoperable with fds and event multiplexer but are not certifiable and have the buffer problem.

--

== UMessage Mapping

A uProtocol message consists of _UAttributes_ and optional payload. The following sections define how these are mapped to/from iceoryx2 Services.

=== UAttributes

[.specitem,oft-sid="dsn~up-transport-iceoryx2-attributes-mapping~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--

> The value of a iceoryx2 Service that is used to convey a uProtocol message *MUST* have a https://docs.rs/iceoryx2/0.6.1/iceoryx2/service/builder/publish_subscribe/struct.Builder.html#method.user_header[user header] as defined in <<uAttributes Mapping to iceoryx2 custom header>>.

Reviewer: ```suggestion
The value of a iceoryx2 Service that is used to convey a uProtocol message *MUST* have a https://docs.rs/iceoryx2/0.6.1/iceoryx2/service/builder/publish_subscribe/struct.Builder.html#method.user_header[user header] as defined in <<uAttributes Mapping to iceoryx2 user header>>.
```

> The value of a iceoryx2 Service that is used to convey a uProtocol message *MUST* have a https://docs.rs/iceoryx2/0.6.1/iceoryx2/service/builder/publish_subscribe/struct.Builder.html#method.user_header[user header] as defined in <<uAttributes Mapping to iceoryx2 custom header>>.

Author: Think we've aligned here.

--

> ==== uAttributes Mapping to iceoryx2 user header

Reviewer: It seems undesirable to be required to do a manual mapping of Rust structs to other Rust structs that are basically the same thing with two additional attributes defined on them :-(
Based on that, I wonder if it wouldn't make more sense to take the same approach as we do in the Zenoh transport, that is, serializing the metadata to its protobuf and add that as (together with the uProtocol major version) into the header?

Author: > This is not correct. iceoryx2 supports dynamic payload, see this example here: https://github.com/eclipse-iceoryx/iceoryx2/tree/main/examples/rust/publish_subscribe_dynamic_data
> 
> Currently, the user can send a strongly typed slices, and with `loan_slice`, you can state how many elements you would like to send at runtime. If the `loan_slice` would exceed the maximum available memory iceoryx2 reallocates memory with a defined `AllocationStrategy`.

Oh! That's pretty cool. Thank you for sharing this example. (And sorry for missing it! 😅)

> It does not yet work with dynamic data which is part of a more complex data type like a struct. So this
> 
> ```rust
> struct MyPayload {
>   data1: RelocatableVec<u64>,
>   data2: RelocatableString
> }
> ```
> 
> Does not yet work but it is on the roadmap.
> 
> But this is a feature for desktop systems or proof-of-concept code. In mission-critical systems this feature would explicitly deactivated and the code would not compile when it is flagged as ASIL-D for instance.

Just to make sure I understood this --
* is it that using loaned slices would not compile for ASIL-D or
* that using the `RelocatableVec` or `RelocatableString` would not compile for ASIL-D

?

Reviewer 2: No, they would compile for ASIL-D but the `AllocationStrategy` cannot be defined in ASIL-D since everything is statically configured and pre-allocated. So if `loan_slice` would exceed the maximum available memory you would get an out-of-memory Err.

Author: Modified to hold serialized UAttributes for simplicity for now.

Author: Let's track this in #291

    Unknown = 2,
    InvalidArgument = 3,
    DeadlineExceeded = 4,
    NotFound = 5,
    AlreadyExists = 6,
    PermissonDenied = 7,
    Unauthenticated = 16,
    ResourceExhausted = 8,
    FailedPrecondition = 9,
    Aborted = 10,
    OutOfRange = 11,
    Unimplemented = 12,
    Internal = 13,
    Unavailable = 14,
    DataLoss = 15,
}
----

==== UPayloadFormat enum for iceoryx2

[source,rust]
----
#[repr(C)]
#[derive(ZeroCopySend)]
enum UPayloadFormat {
    Unspecified = 0,
    ProtoBufWrappedInAny = 1,
    ProtoBuf = 2,
    Json = 3,
    Someip = 4,
    SomeipTlv = 5,
    Raw = 6,
    Text = 7,
    Shm = 8,
}
----

==== Message Priority

TODO(plevasseur@gmail.com): I don't see mention of a way to set priority natively for a Service in the iceoryx2 docs.

=== Payload Mapping

[.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-mapping~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--
An iceoryx2 message that is used to convey a uProtocol message *MUST* contain in its payload the unaltered value of the UMessage's _payload_ field.
--

[.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-mapping~2",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--

> An iceoryx2 message that is used to convey a uProtocol message *MUST* be zero-copy and shared-memory compatible (i.e. be fixed in size, contain no pointers, and so on).

Reviewer: Then I would suggest that we explicitly state here that the payload must implement the `ZeroCopySend` trait. We seem to be talking about Rust based implementations only anyway ...

--

== iceoryx2 Service Name Structure

> Message producers publish messages to the iceory2 inter-process communication using _Service Names_. Other clients can indicate their interest in particular  or patterns in order to receive the messages that are being published using matching keys.

Reviewer: ```suggestion
Message producers publish messages to the iceory2 inter-process communication using _Service Names_. Other clients can indicate their interest in particular Services or patterns in order to receive the messages that are being published using matching names.
```

The iceoryx2 _Service Name_ that is used to transfer a uProtocol message is derived from the message's `source` and `sink` attributes.

[.specitem,oft-sid="dsn~up-transport-iceoryx2-service-name~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]
--

> The ServiceName of an iceoryx2 message containing a _Publish_ UMessage **MUST** consist of the following segments:

Reviewer: is it *ServiceName* or *Service Name*? We should use the term consistently across the spec ...

Author: Tried to find and replace all to _ServiceName_

> `up/[source.authority]/[source.ue_type]/[source.ue_instance]/[source.ue_version]/[source.resource]/{}/{}/{}/{}/{}`

Reviewer: based on a quick glance at the iceoryx2 docs, I am not sure if wildcards are supported in `ServiceName`s ...

Author: Thanks for the feedback @elfenpiff!

> Wildcards are not supported.

These are not supported in the ServiceName and never will be?

---------

Also, to make sure I am clear on how we should be using iceoryx2, I have some questions.

Here's what a uProtocol UUri looks like, I'll reference this:
![image](https://github.com/user-attachments/assets/703304c8-9204-4899-84b3-d93754ed7da0)

(pulled from [this](https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/uri.adoc) other spec page)

### First example

Let's say that I had some "service A" (I use this not in the iceoryx2 definition) which sends two messages representing two different pieces of information `Foo`s and `Bar`s RESOURCE_IDs.

Would I have a _single_ [`Service`](https://docs.rs/iceoryx2/latest/iceoryx2/service/trait.Service.html) for "service A" sending its two messages?

Or _two_ [`Service`](https://docs.rs/iceoryx2/latest/iceoryx2/service/trait.Service.html)s for "service A" sending its two messages?

### Second example

Now imagine that I have two services: "service A" and "service B".

"Service A" sends two messages representing two different pieces of information `Foo`s and `Bar`s RESOURCE_IDs.

And I have some "service B" which wants to subscribe to _all_ resources originating from "serviceA".

Is it possible to do this?

Would that then mean that "service A" should be two [`Service`](https://docs.rs/iceoryx2/latest/iceoryx2/service/trait.Service.html)s to allow for the attaching of attributes to each service indicating that the source UUri is the same except for the RESOURCE_ID?

And then in "service B" it would be possible to iterate over the services, picking out all for which the:
* AUTHORITY_NAME
* ENTITY_ID
* ENTITY_VERSION_MAJOR

matched and then subscribing to them?

Sorry this got quite long 😅 Hope that was clear-ish

Author: Ah I saw a couple more we _might_ be able to tackle via text:

> But if you want to be able to communication `X` and `T` over the same service, you could create a sum type (enum in Rust, variant in C++) and use it as payload.
> 
> ```rust
> pub enum MyPayload {
>   X(X),
>   Y(Y)
> }
> ```

This would enable one or the other type to be sent, but if we wanted to subscribe specifically to one resource out of a possible two, we'd have to follow your advice here, right?

> * Second example:
> 
> > Would that then mean that "service A" should be two [Service]>(https://docs.rs/iceoryx2/latest/iceoryx2/service/trait.Service.html)s to allow for the attaching of attributes to each service indicating that the source UUri is the same except for the RESOURCE_ID?
> > And then in "service B" it would be possible to iterate over the services, picking out all for which the:
> > AUTHORITY_NAME
> > ENTITY_ID
> > ENTITY_VERSION_MAJOR
> > matched and then subscribing to them?
> 
> This is exactly how you would realize it.

Author: > But maybe it is enough to just support * and then you could write My*Service*Fuu. But even this could pose some interesting challenges especially for certification - but nothing we couldn't handle.

For example, the Zenoh UTransport spec sections [4.1. UUri Encoding Rules](https://github.com/eclipse-uprotocol/up-spec/blob/main/up-l1/zenoh.adoc#41-uuri-encoding-rules) and [4.2. Examples](https://github.com/eclipse-uprotocol/up-spec/blob/main/up-l1/zenoh.adoc#42-examples) shows a bit of what we'd aim to achieve.

Bits like this:

```
up://*/FFFFFFFF/FF/FFFF
```

or

```
up://*/FFFF05A1/2/FFFF
```

I suppose this may be a bit "simpler" alternative to the approach of using Attributes, but what do you think?

(again -- if these are too much to answer here we can wait till we meet up!)

Reviewer 2: > This would enable one or the other type to be sent, but if we wanted to subscribe specifically to one resource out of a possible two, we'd have to follow your advice here, right?

Yes

An iceoryx2 message that is used to convey a uProtocol message *MUST* be zero-copy and shared-memory compatible:

* is self contained, no heap, no pointers to external sources, no handles to external resources (files etc)
* has a uniform memory representation
* does not use pointers to manage their internal structure
* has a lifetime of the entire program duration
--

[.specitem,oft-sid="dsn~up-transport-iceoryx2-payload-zero-copy-rust~1",oft-needs="impl,utest",oft-tags="TransportLayerImpl"]

> --

Reviewer: We can leave it like that but I would suggest to remove the specitem declaration and just leave it as a hint. Otherwise I won't be able to cover that specitem if I implement the transport in C++, right?

Author: Ah... leave it as a hint => no annotation needed for OFT at all then? That's doable, just confirming.

Author: Updated in [8eb51ed](https://github.com/eclipse-uprotocol/up-spec/pull/285/commits/8eb51ed633b7cc6a9fc3cf95f6671eb31c7bdec1). Should this work?
